rust/azure_iot_operations_services/tests/azure_device_registry_client_tests.rs

```diff
@@ -0,0 +1,970 @@
+                    },
+                    TIMEOUT,
+                )
+                .await
+                .unwrap();
+            log::info!(
+                "[{log_identifier}] Updated device response after unobserve: {response_after_unobs:?}",
+            );
+            // wait for the receive_notifications_task to finish to ensure any failed asserts are captured.
+            assert!(receive_notifications_task.await.is_ok());
+
+            // Shutdown adr client and underlying resources
+            assert!(azure_device_registry_client.shutdown().await.is_ok());
+
+            exit_handle.try_exit().await.unwrap();
+        }
+    });
+
+    assert!(
+        tokio::try_join!(
+            async move { test_task.await.map_err(|e| { e.to_string() }) },
+            async move { session.run().await.map_err(|e| { e.to_string() }) }
+        )
+        .is_ok()
+    );
+}
+
+#[tokio::test]
+async fn observe_device_notify() {
+    let log_identifier = "observe_device_update_notifications_network_tests-rust";
+    if !setup_test(log_identifier) {
+        return;
+    }
+    let (session, azure_device_registry_client, exit_handle) =
+        initialize_client(&format!("{log_identifier}-client"));
+
+    let test_task = tokio::task::spawn({
+        async move {
+            let mut observation = azure_device_registry_client
+                .observe_device_update_notifications(
+                    DEVICE1.to_string(),
+                    ENDPOINT1.to_string(),
+                    TIMEOUT,
+                )
+                .await
+                .unwrap();
+            log::info!("[{log_identifier}] Device update observation: {observation:?}");
+
+            let first_notification_notify = Arc::new(Notify::new());
+            let cleanup_notify = Arc::new(Notify::new());
+            let receive_notifications_task = tokio::task::spawn({
+                let first_notification_notify = first_notification_notify.clone();
+                let cleanup_notify = cleanup_notify.clone();
+
+                async move {
+                    log::info!("[{log_identifier}] Device update notification receiver started.");
+                    let mut count = 0;
+                    let mut first_notification_sent = false;
+                    loop {
+                        tokio::select! {
+                            notification_result = observation.recv_notification() => {
+                                if let Some((device, _)) = notification_result {
+                                    count += 1;
+                                    log::info!("[{log_identifier}] Device Observation received: {device:?}");
+
+                                    if count == 1 {
+                                        // Signal that we got the first notification
+                                        if !first_notification_sent {
+                                            first_notification_notify.notify_one();
```

Comment:
not sure if you want to verify that the update notification is the one you sent before calling notify_one(), just to make sure you aren't getting some notification from a different test run or something weird